polaris-react/src/components/Scrollable/Scrollable.tsx

```diff
@@ -40,218 +38,95 @@ export interface ScrollableProps extends React.HTMLProps<HTMLDivElement> {
-      scrollHeight - scrollTop <= clientHeight + LOW_RES_BUFFER;
-
-    if (canScroll && hasScrolledToBottom && onScrolledToBottom) {
-      onScrolledToBottom();
-    }
-
-    this.setState({
-      topShadow: shouldTopShadow,
-      bottomShadow: shouldBottomShadow,
-      scrollPosition: scrollTop,
-      canScroll,
-    });
-  };
 
-  private scrollHint = () => {
-    const {scrollArea} = this;
-    if (scrollArea == null) {
-      return;
-    }
-    const {clientHeight, scrollHeight} = scrollArea;
-    if (
-      PREFERS_REDUCED_MOTION ||
-      this.state.scrollPosition > 0 ||
-      scrollHeight <= clientHeight
-    ) {
-      return;
-    }
+    const handleScroll = () => {
+      requestAnimationFrame(() => {
+        const {scrollTop, clientHeight, scrollHeight} = currentScrollArea;
+        const isBelowTopOfScroll = Boolean(scrollTop > 0);
+        const isAtBottomOfScroll = Boolean(
+          scrollTop + clientHeight >= scrollHeight - LOW_RES_BUFFER,
+        );
 
-    const scrollDistance = scrollHeight - clientHeight;
-    this.toggleLock();
-    this.setState(
-      {
-        scrollPosition:
-          scrollDistance > MAX_SCROLL_DISTANCE
-            ? MAX_SCROLL_DISTANCE
-            : scrollDistance,
-      },
-      () => {
-        window.requestAnimationFrame(this.scrollStep);
-      },
-    );
-  };
+        setTopShadow(isBelowTopOfScroll);
+        setBottomShadow(!isAtBottomOfScroll);
 
-  private scrollStep = () => {
-    this.setState(
-      ({scrollPosition}) => {
-        const delta = scrollPosition * DELTA_PERCENTAGE;
-        return {
-          scrollPosition: delta < DELTA_THRESHOLD ? 0 : scrollPosition - delta,
-        };
-      },
-      () => {
-        if (this.state.scrollPosition > 0) {
-          window.requestAnimationFrame(this.scrollStep);
-        } else {
-          this.toggleLock(false);
+        if (isAtBottomOfScroll && onScrolledToBottom) {
+          onScrolledToBottom();
         }
-      },
-    );
-  };
-
-  private toggleLock(shouldLock = true) {
-    const {scrollArea} = this;
-    if (scrollArea == null) {
-      return;
-    }
-
-    EVENTS_TO_LOCK.forEach((eventName) => {
-      if (shouldLock) {
-        scrollArea.addEventListener(eventName, prevent);
-      } else {
-        scrollArea.removeEventListener(eventName, prevent);
-      }
-    });
-  }
-
-  private scrollToPosition = (scrollY: number) => {
-    this.setState({scrollPosition: scrollY});
-  };
-}
+      });
+    };
+
+    const handleResize = debounce(handleScroll, 50, {trailing: true});
+
+    stickyManager.current?.setContainer(currentScrollArea);
+    currentScrollArea.addEventListener('scroll', handleScroll);
+    globalThis.addEventListener('resize', handleResize);
+
+    handleScroll();
+
+    return () => {
+      currentScrollArea.removeEventListener('scroll', handleScroll);
+      globalThis.removeEventListener('resize', handleResize);
+    };
+  }, [stickyManager, onScrolledToBottom]);
```

Comment:
One concern I have is that if `onScrolledToBottom` isn't memoized and the parent re-renders frequently on scroll, we'll be polluted with event listeners and `onScrolledToBottom` will be called multiple times as each debounce will be unique. Something we'll want to keep an eye on when rolling this out 👀

Reply:
Great eye. I wonder if this could trigger cyclical renders. 🤔 I'm gonna look at this a little closer before shipping.